### ISSUE TITLE
Idea: 6309 optimizations?

### ISSUE THREAD

**Issue by drencorxeen** [MyTDT-Mysoft/COCO-FastLoader#3]
You know I was thinking last night and was wondering how much faster could it be made using any of the HD63C09E's enhanced instructions for this?

Some some random thing that came to mind. :)

**Comment by MyTDT-Mysoft** [comment]
well... there's already some "time left" that i just wait to sync... the
problem is the frequency... even the 88khz barely works :P now imagine if
we try a 132khz or 176khz hehe

Em sex., 1 de jan. de 2021 às 14:41, David Ladd <notifications@github.com>
escreveu:

**Comment by MyTDT-Mysoft** [comment]
i'm actually using <14700 bits/s , and so i encode it as 14700*3 = 44100hz so i read 3 samples... and get the one that reads two times... (so that the audio does not need to be in EXACT sync), but on the program i'm actually reading 5 samples theres... and get the one that have 3... 

but so that said... 3 is enough... and so its possible then to squeeze the rest of the program to work at that speed, which means even without double clock... the theorical maximum frequency could be 24500*3 = 73500hz... which would boost the minimum speed from 1600 bytes/s to 2666 bytes/s... and with double clock you have 147000hz and up to 5333 byte/s... 

so yes i tried to keep the frequency to a standard to ensure that the player or sound card... would work correclty... and i dont know if the coco circuitry can do the 147000hz either way

also... i use a stop bit... so after sending 8 bits... it always send a 9 bit that is opposite to the 8bit, that to ensure a resync in case frequencies are slight off on either side... and to avoid the capacitor to not keep the charge like if you would send a lot of 0's or lot of 1's on my tests... coco could barely keep up with 15 consecutive 0's or 1's... so the 9, is already close enough, altough i guess with double frequency those 15 becomes 30...  and thats why the actual byte rate is 14700/9 = ~1633 bytes.

also after the sync flip... of 9 bit.. you have the cycles for the remaining two "sample" that are not required now, to do the bookeping, which on fast load, is just move the byte to desination incrementing the index, and compare if the block is over... after each block... the algorithm does a resync (wait for a pattern that discharges and full charges the capacitors more or less), whose delay i believe equivalent of 3 bytes. but the program sends 12? bytes, i.e. the sync 4 times... so that the program during the block... can do more bookeping like having some good hundred cycles to do whatever... like check keyboard... but the program dont do much there... the STATS that can be seen on screen are written by the loader sending a block of 32bytes to the screen area every second or so... i dont remember which but the overhead is not that significant anyway... which would reduce the rate from ~1633bytes/s to the ~1600bytes/s hehe

but yes... i could attempt to test a high speed version without the double clock, and with the 6309 is indeed possible to have the 88khz version without double clock, and so that said a 6309 with double clock... and 176khz would be indeed possible, but i dont have a 6309 setup to try hehe... and i think 176khz is too close to the limit of 192khz of my Intel HD video card for to be to accurate enough hehe

on an alternative setting....  maybe using the same strategy that CLOAD does... would probabily work as well... because of the niquist rule... as with 44khz... it would do a freq of 22khz, for the 0's and 11khz for 1's, which is... which is ~16537 bauds (average bit second), instead of the fixed 14700 bits/s

**Comment by drencorxeen** [comment]
It was only an idea :)
Just something I was thinking of and figured I would mention it just in case it hadn't come up before.

It also looks like you have already spent a LOT of time figuring out all the math and timing as it is.

I am for sure impressed with what has already been done with this.
Not sure if you saw my comparison video on YouTube or not that I linked on your video, but yeah just the R0 setting is great and I plan on making up some WAV files for writing to Audio CD for tools ready to use when needed for the CoCo :)

I also wonder how well the COCO-FastLoader would work on Dragon systems since most of their software was primarily on cassette.

**Comment by MyTDT-Mysoft** [comment]
yeah... what i need to do next, that was planned but i didnt had time...
was to add option to load .CAS files, .BAS files, and to be able to upload
files from inside .DSK files directly... but yeah not sure if this works on
dragon or not, i assume even the casette format is different isnt it?

Em sex., 1 de jan. de 2021 às 18:18, David Ladd <notifications@github.com>
escreveu:

**Comment by drencorxeen** [comment]
I believe the format is the same, but not sure about the BASIC tokens though.
I know I have loaded ML programs between both without problem.  Only big difference there is the keyboard matrix.
I will see if I can find the Dragon Tano ROM's and try the Dragon Tano system and try the same load with the R0 COCO-FastLoad.

**Comment by drencorxeen** [comment]
Ok I just tried the R0 one and after LOADER finishes loading it switches to a < and then just does nothing after that.
So I guess the Dragon Basic is just enough different that wont let it work without modification. :( darn

**Comment by MyTDT-Mysoft** [comment]
> Ok I just tried the R0 one and after LOADER finishes loading it switches to a < and then just does nothing after that. So I guess the Dragon Basic is just enough different that wont let it work without modification. :( darn

yeah since i use multi-block to do the "auto start", maybe its not even starting with the code... i can test on the emulator to see how far the problem goes, maybe just need a different auto-run routine, because i assume the I/O to read from casette is the same between both, and that the processor timings are equivalent as well... since after all it works on my CP-400 and on original COCO-2, so i suspect the difference may be just on the auto-run code hehe
